src/core/models.py

```diff
@@ -0,0 +1,195 @@
+import logging
+from typing import Literal, MutableMapping
+
+from ops.model import Application, Relation, Unit
+from src.literals import CHARM_USERS
+
+logger = logging.getLogger(__name__)
+
+# TODO: move to some better literals place
+SUBSTRATE = Literal["vm", "k8s"]
+CLIENT_PORT = 2181
+SECURE_CLIENT_PORT = 2182
+SERVER_PORT = 2888
+ELECTION_PORT = 3888
+
+class StateBase:
+    def __init__(self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATE):
+        self.relation = relation
+        self.component = component
+        self.substrate = substrate
+
+    @property
+    def relation_data(self) -> MutableMapping[str, str]:
+        if not self.relation:
+            return {}
+
+        return self.relation.data[self.component]
+
+    def update(self, items: dict[str, str]) -> None:
+        if not self.relation:
+            return
+
+        self.relation_data.update(items)
```

Comment:
I'm a bit lost here, how this would be "written back" to the actual relaion.

In any case my comment applies both to here and to the RO properties below.

We can't use the databag anymore (see https://discourse.canonical.com/t/data-platform-relation-interfaces-using-juju-secrets/2808). So we'll have to make sure that we (can) access relation and manipulate data using corresponding DP-libs interface functions. Not directly.

I wonder now how that may introduce an extra twist on the current structure...